main.go

```diff
@@ -500,6 +506,33 @@ func allUpper(s string) bool {
 	return true
 }
 
+// isHex returns whether all bytes of s are hex digits.
+func isHex(s string) bool {
+	for _, b := range s {
+		b |= 'a' - 'A' // Lower case in the relevant range.
+		if (b < '0' || '9' < b) && (b < 'a' || 'f' < b) {
+			return false
+		}
+	}
+	return true
+}
+
+var (
+	fset = token.NewFileSet()
+	s    scanner.Scanner
+)
+
+// isNumber abuses the go/scanner to check whether word is a number.
+func isNumber(word string) bool {
+	var errored bool
+	eh := func(_ token.Position, _ string) {
+		errored = true
+	}
+	s.Init(fset.AddFile("", fset.Base(), len(word)), []byte(word), eh, 0)
+	_, tok, lit := s.Scan()
+	return !errored && lit == word && (tok == token.INT || tok == token.FLOAT || tok == token.IMAG)
```

Comment:
Hmm. I was initially going to suggest https://pkg.go.dev/go/parser#ParseExpr, but that's potentially more expensive if e.g. the word has special characters in it.

Since you only care about int/float/imag, you could also consider direct calls to the three strconv APIs. That might be simpler, though then you'd have to try each of the three in sequence.